execute/README.md

```diff
@@ -0,0 +1,456 @@
+# Flux Engine Developer Guide
+
+This guide covers the Flux Engine design and related concepts.
+The engine includes:
+
+* Execution engine
+* Builtin sources and transformations
+* Memory layout and resource management
+
+## Concepts
+
+The following concepts are important to understanding the Flux engine and the operations that can be executed within the engine.
+This guide will reference these concepts with the assumption that the reader knows what is contained in this section.
+
+### Tables and Table Chunks
+
+In Flux, a **Table** is a set of ordered rows grouped together by a common **group key**.
+The **group key** is a set of key/value pairs where the value in the column with the same key name matches the value.
+In practical terms, this means that for any columns that are in the group key, the values in that table will all be the same.
+
+The table will also contain columns that are not referenced by the group key.
+These columns are free to be any value as long as they are all the same type.
+
+For a table, the following invariants should hold:
+
+* All values inside a column that is part of the group key should have the same value.
+* That value should match the group key's value.
+* A column with a name should have one and only one type.
+
+The last of these invariants is **loosely** held.
+This means that while they should be true, individual transformations may not make an effort to enforce them.
+That means that if a transformation would have its performance affected by enforcing the invariant and the invariant does not affect the transformation's correctness, the transformation may not enforce it when producing the output.
+For an example, the `group()` transformation does not enforce the last invariant and relies on downstream transformations to enforce it.
+
+The set of rows contained within a table are processed in chunks.
+A table chunk is a subset of the rows within a table that are in-memory.
+
+The rows within a table are **ordered**.
+When table chunks are sent, the rows within the first are all ordered before the rows in the second.
+Transformations may choose to change the order, or they may choose to have the order be meaningful.
+For example, the `derivative()` function gives meaning to the order of the input rows while `sort()` will rearrange the order.
+
+A table chunk is composed of an arrow array for each column.
```

Comment:
Suggestion: "A table chunk is composed of a list of Arrow arrays, each array corresponding to a column of the table.